feat: add `math/base/special/log1pf`

0PrashantYadav0 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   added base implementation for `math/base/special/log1pf`
-   added benchmark, example and test.
-   added docs and C-implementation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   ref #649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add math/special/log1pf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
/stdlib update-copyright-years

0PrashantYadav0 commented:
@gunjjoshi @kgryte Please review this pr as it is important for other packages.

stdlib-bot committed:
chore: update copyright years

0PrashantYadav0 committed:
docs: improve function documentation for stdlib_base_log1pf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,357 @@
+// MAIN //
+
+/**
+* Evaluates the natural logarithm of \\(1+x\\) in single‐precision floating-point number.
```
```suggestion
* Evaluates the natural logarithm of \\(1+x\\) as a single‐precision floating-point number.
```

gunjjoshi replied:
@0PrashantYadav0 Can you change this description at other places too, in this PR?

0PrashantYadav0 replied:
Sure

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,357 @@
+/**
+* Evaluates the natural logarithm of \\(1+x\\) in single‐precision floating-point number.
+*
+* ## Method
```
We don't need this "Method" part, as the [freeBSD](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/s_log1pf.c?revision=367086&view=markup) source which you've referred here doesn't have it. We do have this in the double-precision implementation, i.e., [`log1p`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/log1p/lib/main.js#L76), but that's only because its reference implementation. i.e., netlib had it.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,357 @@
+*
+* ## Notice
+*
+* The following copyright, license, and long comment were part of the original implementation available as part of [FreeBSD]{@link https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/s_log1pf.c}. The implementation follows the original, but has been modified for JavaScript.
```
@0PrashantYadav0 You have mentioned the reference implementation as freeBSD here, but your implementation seems to be varying from it. It seems more like that of the double-precision netlib implementation of `log1p`. Which implementation have you actually followed?

0PrashantYadav0 replied:
I followed freeBSD implementation but also take reference from `stdlib/math/base/special/log1pf`

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

0PrashantYadav0 committed:
chore: changes as per code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
@gunjjoshi I made the changes you asked for. Please review also i use `freeBSD` for my implementation along with the reference of `log1p`

gunjjoshi commented:
@0PrashantYadav0 Thanks for working on this, but after commit https://github.com/stdlib-js/stdlib/pull/5803/commits/46d0f11477bb2d055bfae81891d7994ca25a2a4d, there doesn't seem to be any change in the implementation. You should be strictly following the freeBSD reference, rather than trying to port the double-precision netlib reference.

0PrashantYadav0 commented:
@gunjjoshi I am not sure what you mean will you please explain this to me.

Neerajpathak07 commented:
@0PrashantYadav0 Refer this [`msun`](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/s_log1pf.c?revision=367086&view=markup) implementation for `log1pf`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/log1pf/src/main.c:
```diff
@@ -0,0 +1,213 @@
+static const float LN2_HI = 6.9313812256e-01f; // 0x3f317180
+static const float LN2_LO = 9.0580006145e-06f; // 0x3717f7d1
+
+// sqrt(2)-1:
+static const float SQRT2M1 = 4.142135623e-01f; // 0x3ed413cf
+
+// sqrt(2)/2-1:
+static const float SQRT2HALFM1 = -2.928932188e-01f; // 0xbe95f618
+
+// 2**-15:
+static const float SMALL = 3.051757812e-05f; // 0x38000000
+
+// 2**-24:
+static const float TINY = 5.960464477e-08;
```
According to the `msun` implementation we won't be needing these vars

0PrashantYadav0 commented:
I tried to implement the freeBSD completely but i am not sure why i am getting wrong values as an output.

0PrashantYadav0 commented:
@gunjjoshi I made few changes according to your previous comments. Please review them. Thank you.

0PrashantYadav0 committed:
chore: added new implementation and corrected docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented:
<!-- stdlib-bot-tracking-issue-closure-check -->
:warning: **Tracking Issue Closure Warning** :warning:

I noticed your PR description contains closing keywords ("Resolves", "Closes", or "Fixes") referencing a "Tracking Issue".

**Why this matters:**
Tracking issues should typically remain open until all related sub-issues are completed. GitHub automatically closes issues with such closing keywords when the PR is merged.

**Required action:**
Use "Progresses:" or "Ref:" instead to reference the tracking issue without automatically closing it.

Thank you for your contribution to the project!

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into math-log1pf

anandkaranubc committed:
refactor: update the implementation to follow project conventions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,220 @@
+		hu &= FLOAT32_SIGNIFICAND_MASK;
+
+		// Earlier sqrt(2) checks should be slightly looser than this one, so we never reach this point with k == 0 after deciding to use the correction term...
+		if ( hu < 0x3504f4 ) { // u < sqrt(2)
```
Is that ok to leave the hex value as is, without creating a separate variable? I am not sure what that would be called.

kgryte replied:
Yes, that is fine.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,220 @@
+		if ( k === 0 ) {
+			return f32( f - R );
+		}
+		// eslint-disable-next-line max-len
```
I hope this line is ok.

anandkaranubc replied:
The "eslint" disable line.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/log1pf/lib/main.js:
```diff
@@ -0,0 +1,220 @@
+		return f32( f - f32( hfsq - f32( s * f32( hfsq + R ) ) ) );
+	}
+	// eslint-disable-next-line max-len
+	return f32( f32( k*LN2_HI ) - f32( f32( hfsq - f32( f32( s * f32( hfsq+R ) ) + f32( f32( k*LN2_LO ) + c ) ) ) - f ) );
```
Same for this one.

anandkaranubc committed:
refactor: update the C implementation to follow project conventions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/log1pf/src/main.c:
```diff
@@ -0,0 +1,207 @@
+			}
+			return x - ( x * x * 0.5f );
+		}
+		// cppcheck-suppress knownConditionTrueFalse
+		if ( ( hx > 0 ) || ( hx <= SQRT2HALFM1 ) ) { // sqrt(2)/2- <= 1+x < sqrt(2)+
+			k = 0;
+			f = x;
+			hu = 1;
+		}
+	}
```
Adding the `// cppcheck-suppress knownConditionTrueFalse` comment because I think cppcheck raises a false positive in here.

I believe it's thinking of it as "If a number is positive or a if it's negative" but this check is more of whether the number lies outside of a given range.

reference: https://github.com/freebsd/freebsd-src/blob/adb4901ac9ae6c2ceb4194d139ed3176f6ed5e55/lib/msun/src/s_log1pf.c#L58-L59

kgryte replied:
A bit odd that cppcheck is flagging this.

anandkaranubc committed:
test: use `ulpdiff` for test comparisons

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: code cleanup

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc reviewed (dismissed):
Thanks @0PrashantYadav0 for the valuable contribution! This will unlock a number of other single-precision packages.

kgryte commented on lib/node_modules/@stdlib/math/base/special/log1pf/docs/repl.txt:
```diff
@@ -0,0 +1,35 @@
+
+{{alias}}( x )
+    Evaluates the natural logarithm of `1+x` as a single-precision
+    floating-point number.
```
```suggestion
    Evaluates the natural logarithm of `1+x` as a single-precision floating-
    point number.
```

kgryte committed:
style: adjust line wrapping

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM.
